docs/aggregations.md

```diff
@@ -0,0 +1,178 @@
+true)` annotation. Timeseries types are immutable, and must have an `id`
+field and a `timestamp` field. The `timestamp` is set automatically by
+`graph-node` to the timestamp of the current block; if mappings set this
+field, it is silently overridden when the entity is saved.
+
+Aggregations are declared with an `@aggregation` annotation instead of an
+`@entity` annotation. They must have an `id` field and a `timestamp` field.
+Both fields are set automatically by `graph-node`. The `timestamp` is set to
+the beginning of the time period that that aggregation instance represents,
+for example, to the beginning of the hour for an hourly aggregation. The
+`id` field is set to the `id` of one of the raw data points that went into
+the aggregation. Which one is chosen is not specified and should not be
+relied on.
+
+**TODO**: add a `Timestamp` type and use that for `timestamp`
+
+**TODO**: figure out whether we should just automatically add `id` and
+`timestamp` and have validation just check that these fields don't exist
+
+Aggregations can also contain _dimensions_, which are fields that are not
+aggregated but are used to group the data points. For example, the
+`TokenStats` aggregation below has a `token` field that is used to group the
+data points by token:
+
+```graphql
+# Normal entity
+type Token @entity { .. }
+
+# Raw data points
+type TokenData @entity(timeseries: true) {
+    id: Bytes!
+    timestamp: Int8!
+    token: Token!
+    amount: BigDecimal!
+    priceUSD: BigDecimal!
+}
+
+# Aggregations over TokenData
+type TokenStats @aggregation(intervals: ["hour", "day"], source: "TokenData") {
+  id: Int8!
+  timestamp: Int8!
+  token: Token!
+  totalVolume: BigDecimal! @aggregate(fn: "sum", arg: "amount")
+  priceUSD: BigDecimal! @aggregate(fn: "last", arg: "priceUSD")
+  count: Int8! @aggregate(fn: "count")
+}
+```
+
+Fields in aggregations without the `@aggregate` directive are called
+_dimensions_, and fields with the `@aggregate` directive are called
+_aggregates_. A timeseries type really represents many timeseries, one for
+each combination of values for the dimensions.
+
+**TODO** As written, this supports buckets that start at zero with every new
+hour/day. We also want to support cumulative statistics, i.e., snapshotting
+of time series where a new bucket starts with the values of the previous
+bucket.
+
+**TODO** Since average is a little more complicated to handle for cumulative
+aggregations, and it doesn't seem like it used in practice, we won't
+initially support it. (same for variance, stddev etc.)
+
+**TODO** The timeseries type can be simplified for some situations if
+aggregations can be done over expressions, for example over `priceUSD *
+amount` to track `totalVolumeUSD`
+
+**TODO** It might be necessary to allow `@aggregate` fields that are only
+used for some intervals. We could allow that with syntax like
+`@aggregate(fn: .., arg: .., interval: "day")`
```

Comment:
feels like more complicated aggregations could benefit from a mapping to do them? Essentially providing a (x, x) -> x function to generalise.